include/nuttx/foc/foc.h

```diff
@@ -0,0 +1,158 @@
+/****************************************************************************
+ * include/nuttx/foc/foc.h
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __INCLUDE_NUTTX_DRIVERS_FOC_FOC_H
+#define __INCLUDE_NUTTX_DRIVERS_FOC_FOC_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/compiler.h>
+
+#include <stdbool.h>
+#include <semaphore.h>
+
+#include <fixedmath.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define FOCDUTY_FROM_FLOAT(d)   (ftob16(d))
+#define FOCDUTY_FROM_FIXED16(d) (d)
+
+#define FOCDUTY_TO_FLOAT(d)     (b16tof(d))
+#define FOCDUTY_TO_FIXED16(d)   (d)
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+/* FOC device fault code */
+
+enum foc_fault_e
+{
+  FOC_FAULT_NONE    = (0),      /* No fault */
+  FOC_FAULT_TIMEOUT = (1 << 1), /* Timeout fault */
+  FOC_FAULT_ARCH    = (1 << 2), /* Arch-specific fault */
+  FOC_FAULT_BOARD   = (1 << 3), /* Board-specific fault */
+};
+
+/* Phase current as signed 32-bit integer */
+
+typedef int32_t foc_current_t;
+
+/* Phase duty cycle as unsigned fixed16.
+ * We use range [0.0 to 1.0] so this gives us a 16-bit resolution.
+ */
+
+typedef ub16_t foc_duty_t;
+
+/* FOC device configuration */
+
+struct foc_cfg_s
+{
+  uint32_t pwm_freq;            /* FOC PWM frequency */
+  uint32_t notifier_freq;       /* FOC notifier frequency */
+};
+
+/* Output data from the FOC device */
+
+struct foc_state_s
```

Comment:
Something I would add here, why not also add support for rotor position sensors using this driver interface? Aka encoders? 

The sensorless foc is very common but they lack of performance at very slow speeds since most of observer implementations need a minimal velocity on motor-shaft to behave correctly.

By allowing to use the direct rotor-sensor data from there, the startup routine of foc controller can be disabled, and, more valuable, it would enable using foc at very small speeds. also makes this information very useful for robotics odometry computations.

Reply:
This is intentional. The assumption is to support any possible phase angle sensor: incremental encoders, serial encoders, resolvers and even sensorless observers. It may happen that you will also need to process the phase angle information before delivering it to the controller (eg. estimation of the phase angle with a hall sensor and rotor speed). Doing this on the kernel side causes a total mess and lack of modularity.